Introduce [@bs.unwrap] for inline variant as external arg

ryanartecona opened the pull request:
Closes #1736 

---

This introduces a `[@bs.unwrap]` attribute as described in #1736 which can be added to an argument to an external, if that argument has an inline anonymous variant as its type, and will have the effect of the boxed value being "unwrapped" from the tagged tuple that bucklescript usually creates for variants at codegen callsites.

For example, this ocaml:

```ocaml
external log :
  (
    [ `String of string
    | `Options of < foo : int > Js.t
    ] [@bs.unwrap]
  )
  -> unit = "console.log" [@@bs.val]

let _ = log (`String "foo")
let _ = log (`Options [%bs.obj { foo = 1 }])
```

will compile to this js:

```js
console.log("foo");

console.log({
      foo: 1
    });
```

One intention is that we can guarantee that if the variant is constructed inline at the callsite, as in ``log (`String "foo")``, then the codegen will produce `log("foo")` without any intermediate boxing/unboxing at all. I _think_ I've done this correctly, and have some basic tests, but definitely need a review. In any case, if the arg isn't constructed directly at the callsite, the emitted js will create an intermediate tuple (js array) and the callsite will index into it to get the wrapped arg, as in:

```js
var arg_string = /* `String */[
  -976970511,
  "hi runtime"
];
console.log(arg_string[1]);
```

I don't know if there are other opportunities to avoid this intermediate boxing that I haven't taken, but none were obvious to me.

As a side note, hi 👋 ! This is my first PR to bucklescript and first time writing OCaml in a little while, so please pick my code apart and let me know what other checks/docs/tests/etc. should be included.

#### TODO

- [x] Document `[@bs.unwrap]` usage & examples
- [ ] Decide on name:  `[@bs.unwrap]` or `[@bs.unbox]`?

bobzhang commented:
are you aware that it is already achievable in current context?

```
type _ x = 
| S :  string x 
| F : (int -> unit) x 
external log :  ('a x [@bs.ignore]) -> 'a -> ..= 
```
The above is a bit verbose, but it is less  ad-hoc

saschatimme commented:
I think @rrdelaney would love this, since he already has a similar more "hackish" solution in [ReasonablyTyped](https://github.com/rrdelaney/ReasonablyTyped).

bobzhang commented on jscomp/core/js_of_lam_variant.ml:
```diff
@@ -84,3 +84,13 @@ let eval_as_int (arg : J.expression) (dispatches : (int * int) list ) : E.t  =
+let eval_as_unwrap (arg : J.expression) : E.t =
+  if arg == E.undefined then
+    E.undefined
+  else
```
this branch is unnecessary? since it is guaranteed to be an array

bobzhang replied:
1. I am not sure if it is needed, can you find an example?
2. You need make sure in the front end such case does not happen
```ocaml
[ `a of int | `b]
```

bobzhang replied:
can you add a ffi failed test like this
https://github.com/BuckleScript/bucklescript/blob/master/jscomp/ounit_tests/ounit_ffi_error_debug_test.ml#L33
run `make snapshotml && make test`

bobzhang replied:
how would you generate code for such case?
```ocaml
external f : ?x:[`a of int] -> int = "" [@@bs.val]
let u = f
```
I suppose it would generate code like this
```js
function u (prim){
   return f(prim?prim[0] [1] : undefined)
}
```
if so can you add it

bobzhang commented on jscomp/syntax/ast_external_attributes.ml:
```diff
@@ -128,8 +128,35 @@ let get_arg_type ~nolabel optional
     | (`Int, _), _ -> Bs_syntaxerr.err ptyp.ptyp_loc Invalid_bs_int_type
+    | (`Unwrap, ptyp_attributes), (Ptyp_variant (row_fields, Closed, _) as ptyp_desc) ->
+      let validity =
+        (List.fold_left
```
can you make it a stand alone function and with some comments which are valid , which are invalid cases?

ryanartecona replied:
done in 9703061

bobzhang commented on jscomp/syntax/ast_external_attributes.ml:
```diff
@@ -494,6 +521,9 @@ let handle_attributes
                          "bs.obj label %s does not support such arg type" name
+                     | Unwrap ->
+                       Location.raise_errorf ~loc
+                         "bs.obj label %s does not support [@bs.unwrap] arguments" name
```
There is no technical reasons it should not be supported, correct?

bobzhang replied:
correct

bobzhang commented on jscomp/test/bs_unwrap_test.ml:
```diff
@@ -0,0 +1,49 @@
+external log1 :
+  (
+    [ `Pair of string * int
+    | `Int of int
```
can you also add some tests for label types, and optional types?

bobzhang replied:
I mean tests like this
```
~f:([`pair of int] [@bs.unwrap])
?f:([`pair of int] [@bs.unwrap])
```
maybe the name `bs.unbox` makes more sense

ryanartecona replied:
In that case there is a test for both below. 

I don't have a strong preference between "unwrap" and "unbox" as the term we use here.

glennsl replied:
FWIW, `unbox`is the easier term to relate to for me, but my impression is that `unwrap` is the more established term, and is used with rust's `Option` type for example: https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap

ryanartecona commented:
> are you aware that it is already achievable in current context?

Yeah. I worked on this feature mostly because it looked like a good first project to get familiar with bs internals, and only secondarily because I though it would be useful. But I do think it's useful!

There's already some discussion in #1736 on the merits of this feature, but I'll add my own take. When writing js and calling a function which can accept, say, a string or an options object in the same argument position, I _think_ about that object conceptually as a variant. So something like ``log (`String "message")`` matches my intuition closer than `log String "message"`.

I think it gives a smoother path for js bindings, especially when automatically generated e.g. via ReasonablyTyped. Without `[@bs.unwrap]`, bindings would have to define a GADT for each unique combination of argument types (and if automatically generated, would also have to be deduped and given a generated name like `union_of_options_or_string`), and users of the library need to be either taught how to use a GADT or shown an example and told not to worry about it. Or, alternatively, you can define a normal variant for each of those argument types and write a helper that does the kind of runtime unwrapping that `[@bs.unwrap]` does here, but which can never be compiled away and still has the effect of each variant argument to an external binding incurring up to 2 extra type definitions an an extra external helper for the runtime unwrapping (which is [what ReasonablyTyped currently does](https://github.com/rrdelaney/ReasonablyTyped/blob/master/src/__tests__/fixtures/union-type.re)).

ryanartecona commented:
FYI: I've rebased this on top of current master, and added user docs for `[@bs.unwrap]` in cc5dbfb6b4339284966a7294b281e54abd77d7c0. I think the last remaining questions are whether all the right tests are in place, whether the docs are up to snuff, and whether to name this thing `[@bs.unwrap]` vs. `[@bs.unbox]`? Anything else?

bobzhang commented on jscomp/test/bs_unwrap_test.ml:
```diff
@@ -0,0 +1,48 @@
+
+let _ = log3 ~req:(`Int 1) ()
+let _ = log3 ~req:(`Int 1) ~opt:(`String "hi") ()
+
```
can you also add a test case as below:
```ocaml
log3 ~req: (`Int 1 ) ?opt:(Some (`String "hi")) ()
log3 ~req:(`Int 1) ?opt:None ()
```

ryanartecona commented:
> I am not sure if it is needed, can you find an example?

You were right that the explicit check for `undefined` in `eval_as_unwrap` was unnecessary. It's removed.

> can you add a ffi failed test

Done in 3f934ea. Though, doing a `make snapshotml` didn't work for me, and it looks like `make force-snapshotml` _did_ work but also made a lot more changes than I expected that I don't understand. Does that commit look ok to you?

> how would you generate code for such case?

I hadn't considered this, but it seems to be working well now. Specifically, this ocaml

```ocaml
external f : ?x:([`a of int] [@bs.unwrap]) -> int = "" [@@bs.val]
let u = f
```

generates exactly this js

```js
function u(prim) {
  return f(prim ? prim[0][1] : undefined);
}
```

> can you also add a test case as below:
> ```
> log3 ~req: (`Int 1 ) ?opt:(Some (`String "hi")) ()
> log3 ~req:(`Int 1) ?opt:None ()
> ```

Oh! I thought that was possible in ocaml, but I can never remember the syntax, that the `~` changes to `?`. I've added some tests for various combinations of those, and also for a case where the optional argument is not a simple side-effect-free expression. Everything is in `jscomp/test/bs_unwrap_test.{ml,js}`.

bobzhang commented:
hi @ryanartecona the failure is probably due to the complexity of git submodule, I am going to replace it with subtree.
The current contribution bar is indeed quite high (we had a complex setup), we are improving it based on existing contributions, sorry for the inconvenience

ryanartecona commented:
@bobzhang Thanks. The build is fixed now.

The problem was that I had done `make force-snapshotml` without having done `cd ocaml; ./configure; make world.opt`. I didn't see that necessity mentioned in the docs somewhere, but that seems to be what fixed it for me, I believe because the bucklescript tests wanted to find a file at `ocaml/parsing/parser.ml` which hadn't been generated locally for me from its `.mly` file.

Separately, it looks like my `make snapshotml` generated a bunch of possibly unwanted changes? A lot of stuff in `jscomp/bin/{bsdep,bspp,whole_compiler}.ml` have changes that look like this:

```diff
modified   jscomp/bin/whole_compiler.ml
@@ -17096,7 +17096,7 @@ val list_variables : Format.formatter -> unit
 
 end = struct
 #1 "lexer.ml"
-# 15 "parsing/lexer.mll"
+# 15 "../ocaml/parsing/lexer.mll"
  
 open Lexing
 open Misc
```

If that's expected or acceptable, no worries from me! I just don't know if it's wrong or bad or what.

No worries on the high bar of contribution. I'm frankly amazed so much has been achieved so far with so few contributors! Thanks for the help, and I'd be happy to test out contributor documentation/workflow improvements in the future if that would be useful.

bobzhang commented:
Hi @ryanartecona, 
>The problem was that I had done make force-snapshotml without having done cd ocaml; ./configure; make world.opt. I didn't see that necessity mentioned in the docs somewhere, but that seems to be what fixed it for me, I believe because the bucklescript tests wanted to find a file at ocaml/parsing/parser.ml which hadn't been generated locally for me from its .mly file.


recently we used git subtree, so that configure is no longer needed: see 
https://github.com/BuckleScript/bucklescript/blob/master/vendor/ocaml/parsing/parser.ml 
in that case, the diff should not be there
Also can you sign the commit message (see CONTRIBUTION.md)  and update Changes.md? thanks

ryanartecona commented:
Ok, I've rebased and rerun `make snapshotml`, and that part of the diff is indeed gone.

I've been sure to sign off on all my commits, and I've added the changelog entry. I think we're good to go?

bobzhang commented:
Thanks!

rrdelaney commented:
Awesome! I'm really looking forward to implementing this in ReasonablyTyped
